### ISSUE TITLE
Regexp testing

### ISSUE THREAD

**Issue by PXshadow** [go2hx/go2hx#162]
# Tests
* [x] TestGoodCompile
* [ ] TestBadCompile
* [x] TestMatch
* [x] TestMatchFunction
* [x] TestCopyMatch
* [x] TestReplaceAll
* [x] TestReplaceAllLiteral
* [x] TestReplaceAllFunc
* [ ] TestQuoteMeta
* [ ] TestLiteralPrefix
* [ ] TestSubexp
* [ ] TestSplit
* [ ] TestParseAndCompile
* [ ] TestOnePassCutoff
* [ ] TestSwitchBacktrack
* [ ] TestDeepEqual
* [ ] TestMinInputLen
* [ ] TestRE2Exhaustive
* [ ] TestRE2Search
* [ ] TestFowler
* [ ] TestLongest
* [ ] TestProgramTooLongForBacktrack
* [ ] TestFind
* [ ] TestFindString
* [ ] TestFindIndex
* [ ] TestFindStringIndex
* [ ] TestFindReaderIndex
* [ ] TestFindAll
* [ ] TestFindAllString
* [ ] TestFindAllIndex
* [ ] TestFindAllStringIndex
* [ ] TestFindSubmatch
* [ ] TestFindStringSubmatch
* [ ] TestFindSubmatchIndex
* [ ] TestFindStringSubmatchIndex
* [ ] TestFindReaderSubmatchIndex
* [ ] TestFindAllSubmatch
* [ ] TestFindAllStringSubmatch
* [ ] TestFindAllSubmatchIndex
* [ ] TestFindAllStringSubmatchIndex
* [ ] TestMergeRuneSet
* [ ] TestCompileOnePass
* [ ] TestRunOnePass

## regexp/syntax

* [x] TestParseSimple
* [x] TestParseFoldCase
* [x] TestParseLiteral
* [x] TestParseMatchNL
* [x] TestParseNoMatchNL
* [x] TestFoldConstants
* [x] TestAppendRangeCollapse
* [ ] TestParseInvalidRegexps
* [ ] TestToStringEquivalentParse
* [ ] TestCompile
* [ ] TestSimplify

# Outline

This is the 2nd time this style of testing issue has been created the first being Fmt testing (#153). The approach seems to work well and helps clearly work through the next blocking error/bug etc. It's really nice to get some helpful advice when issues arise and makes it easier to work on these type of long term marathon test suites that can last multiple months. The one big thing I will change is I will document much faster when the issue is no longer blocking so that no time is wasted on the part of @elliott5 and potential future contributors.

# Completion goals

Unblock progress on Fmt testing (#153) and continue the completion goals for that test suite. A secondary goal is to get parsing style code working more rigorously in go2hx, that will help open the door for json and other formats.

**Comment by PXshadow** [comment]
Infinite loop for the last test in TestBadCompile here is the test isolated:
```haxe
package main

import (
	"fmt"
	"regexp"
	"strings"
)

type stringError struct {
	re  string
	err string
}

func main() {
	t := stringError{strings.Repeat(`\pL`, 27000), "expression too large"}
	re, err := regexp.Compile(t.re)
	fmt.Println(re, err)
}
```

I traced it to getting stuck in this [loop](https://github.com/golang/go/blob/master/src/regexp/syntax/parse.go#L926).

**Comment by elliott5** [comment]
The most suspect part of this code to me are the 7
```go
	break BigSwitch
```
statements, I'd look there first.

**Comment by elliott5** [comment]
Also both of the switch statements have the `default` branch as the first one:

https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/regexp/syntax/parse.go#L897

https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/regexp/syntax/parse.go#L930

**Comment by PXshadow** [comment]
@elliott5 Good guess, I swapped the default branch for both switches to be the last case, but unfortunately it didn't change the hanging forever issue.

I agree with you the break jumps are almost guaranteed to be the problem, the issue is I need to try and isolate the problem somehow, as it's hard to read such a large macro code generation of a label/goto control flow. I guess I'll just trial and error and see what changes until I get something.

**Comment by elliott5** [comment]
I suggest adding a test setup something like:

```go
    bigSwitchDebug := ""
bigSwitch:
    // you could print it out for logging
   bigSwitchDebug = ""

    // ... in each of the 7 locations...
    if bigSwitchDebug != "" {
        panic( "last break bigSwitch at " + bigSwitchDebug +" did not work" )
    }
    bigSwitchDebug = "some name to identify where we are now"
    break bigSwitch
```

**Comment by PXshadow** [comment]
Great idea will do @elliott5 !

**Comment by PXshadow** [comment]
I tried deleting all of the break bigSwitch(s) and I found that the Haxe code still iterated over the big while loop in parse 2 more times then Go, and it lead to a change in this [functions](https://github.com/golang/go/blob/70f98a251efdbfd619c4ff466a43da299ad04752/src/regexp/syntax/parse.go#L1453) behavior, and it also uses default switch as 0 case and label breaking:

```go
Switch:
	switch c {
	default:
		if c < utf8.RuneSelf && !isalnum(c) {
			// Escaped non-word characters are always themselves.
			// PCRE is not quite so rigorous: it accepts things like
			// \q, but we don't. We once rejected \_, but too many
			// programs and people insist on using it, so allow \_.
			return c, t, nil
		}
```

I am tempted in either rewriting the controlFlow macro system or manual writing tests to break it.

**Comment by PXshadow** [comment]
The functionality of parseEscape is the same, the difference is right after, it's not jumping out of all of the loops needed I believe, so it keeps re-looping.

**Comment by elliott5** [comment]
Getting the control flow right is central to the success of the project overall ... so you can't spend too much time on it.

If the generated code needs to be uglier in order to make the control flow work reliably, that is a price worth paying in the short-term.

**Comment by PXshadow** [comment]
I agree with you @elliott5 , I suspect it's a small minor difference now.

**Comment by PXshadow** [comment]
Well! I was kind of right @elliott5 it's so incredibly slow, that I thought, there was no way the control flow was working, but I'm tracing iterations now and it's running, just really really really slowly. I feel very silly that I did not expect this sooner, I just thought it was a control flow problem so I only spent time looking in that direction, and because tt wasn't it would've taken forever.

I sense a pattern here. Very non optimized code, inside of large iteration blocks creates a bad combination.

By the way if you want to have fun profiling this type of code on mac, you can use the [new mac profiler](https://github.com/HaxeFoundation/hashlink/pull/588) thanks to @onehundredfeet 's work on it.

You can easily use it with the ``rnd`` test bed by running:
```sh
haxe rnd.hxml -D profile
```
Then after the run, use the chrome debugger as documented in the [Hashlink wiki to display the results](https://github.com/HaxeFoundation/hashlink/wiki/Profiler#displaying-results).

**Comment by PXshadow** [comment]
<img width="205" alt="Screen Shot 2023-03-14 at 10 31 34" src="https://user-images.githubusercontent.com/9813078/224957778-43662062-ccf5-45df-8037-71939ecd2e4b.png">

I think this is a similar issue to Slice before, where capacity and resizing is not being handled in a similar way as Go and therefore there is a high performance penalty for otherwise avoided reallocation, because of preemptive growing, rather then a new allocation for every new element.

**Comment by PXshadow** [comment]
New failing test for regexp/syntax:

go2hx:
```go
cat{eol{}eol{}eol{}eol{}}
```

go:
```go
cat{eol{}eol{}}
```

```
package main

import (
	"fmt"
	"regexp/syntax"
	"strings"
	"unicode"
)

func main() {
	r, err := syntax.Parse(`$$`, syntax.Flags(204))
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(dump(r))
	}
}

// dump prints a string representation of the regexp showing
// the structure explicitly.
func dump(re *syntax.Regexp) string {
	var b strings.Builder
	dumpRegexp(&b, re)
	return b.String()
}

var opNames = []string{
	syntax.OpNoMatch:        "no",
	syntax.OpEmptyMatch:     "emp",
	syntax.OpLiteral:        "lit",
	syntax.OpCharClass:      "cc",
	syntax.OpAnyCharNotNL:   "dnl",
	syntax.OpAnyChar:        "dot",
	syntax.OpBeginLine:      "bol",
	syntax.OpEndLine:        "eol",
	syntax.OpBeginText:      "bot",
	syntax.OpEndText:        "eot",
	syntax.OpWordBoundary:   "wb",
	syntax.OpNoWordBoundary: "nwb",
	syntax.OpCapture:        "cap",
	syntax.OpStar:           "star",
	syntax.OpPlus:           "plus",
	syntax.OpQuest:          "que",
	syntax.OpRepeat:         "rep",
	syntax.OpConcat:         "cat",
	syntax.OpAlternate:      "alt",
}

// dumpRegexp writes an encoding of the syntax tree for the regexp re to b.
// It is used during testing to distinguish between parses that might print
// the same using re's String method.
func dumpRegexp(b *strings.Builder, re *syntax.Regexp) {
	if int(re.Op) >= len(opNames) || opNames[re.Op] == "" {
		fmt.Fprintf(b, "op%d", re.Op)
	} else {
		switch re.Op {
		default:
			b.WriteString(opNames[re.Op])
		case syntax.OpStar, syntax.OpPlus, syntax.OpQuest, syntax.OpRepeat:
			if re.Flags&syntax.NonGreedy != 0 {
				b.WriteByte('n')
			}
			b.WriteString(opNames[re.Op])
		case syntax.OpLiteral:
			if len(re.Rune) > 1 {
				b.WriteString("str")
			} else {
				b.WriteString("lit")
			}
			if re.Flags&syntax.FoldCase != 0 {
				for _, r := range re.Rune {
					if unicode.SimpleFold(r) != r {
						b.WriteString("fold")
						break
					}
				}
			}
		}
	}
	b.WriteByte('{')
	switch re.Op {
	case syntax.OpEndText:
		if re.Flags&syntax.WasDollar == 0 {
			b.WriteString(`\z`)
		}
	case syntax.OpLiteral:
		for _, r := range re.Rune {
			b.WriteRune(r)
		}
	case syntax.OpConcat, syntax.OpAlternate:
		for _, sub := range re.Sub {
			dumpRegexp(b, sub)
		}
	case syntax.OpStar, syntax.OpPlus, syntax.OpQuest:
		dumpRegexp(b, re.Sub[0])
	case syntax.OpRepeat:
		fmt.Fprintf(b, "%d,%d ", re.Min, re.Max)
		dumpRegexp(b, re.Sub[0])
	case syntax.OpCapture:
		if re.Name != "" {
			b.WriteString(re.Name)
			b.WriteByte(':')
		}
		dumpRegexp(b, re.Sub[0])
	case syntax.OpCharClass:
		sep := ""
		for i := 0; i < len(re.Rune); i += 2 {
			b.WriteString(sep)
			sep = " "
			lo, hi := re.Rune[i], re.Rune[i+1]
			if lo == hi {
				fmt.Fprintf(b, "%#x", lo)
			} else {
				fmt.Fprintf(b, "%#x-%#x", lo, hi)
			}
		}
	}
	b.WriteByte('}')
}
```

**Comment by elliott5** [comment]
Immediate reaction is to wonder if Haxe creates end-of-line as `<carrage-return><newline>` (windows style) rather than `<newline>` (unix style) - which would explain why two `eol{}` tokens appear for each `$`.

This is pure speculation on my part!

**Comment by PXshadow** [comment]
Good guess, I found the problem however, it's to do with the AST generation where $ sign gets a corresponding $ sign in order to remove string Interpolation, however if the string is not a single parenthesis string this adds an unnecessary $, which is happening now.

**Comment by PXshadow** [comment]
New problem regex/syntax:

### hl
```sh
=== RUN  TestParseInvalidRegexps
Exception: Can't cast stdgo.regexp.syntax.ErrorCode_asInterface to stdgo._GoString.GoStringData
```
### interp
```sh
=== RUN  TestParseInvalidRegexps
Exception: Invalid string
Called from stdgo.regexp.syntax._Syntax.Syntax_Fields_.testParseInvalidRegexps (./stdgo/regexp/syntax/Syntax.hx line 2721 column 31)
Called from stdgo.regexp.syntax._Syntax.Syntax_Fields_.parse (./stdgo/regexp/syntax/Syntax.hx line 1727 column 16)
Called from stdgo.regexp.syntax._Syntax.Syntax_Fields_._parse (./stdgo/regexp/syntax/Syntax.hx line 2119 column 21)
Called from local function #1 (./stdgo/regexp/syntax/Syntax.hx line 1753 column 17)
Called from local function #1 (./stdgo/regexp/syntax/Syntax.hx line 1748 column 33)
```

**Comment by elliott5** [comment]
The offending code is at https://github.com/go2hx/go2hx/blob/ee3a8ced204635456c6df3df2c1f67893ff39ad1/stdgo/regexp/syntax/Syntax.hx#L1748

`_r` is set in an earlier function with the code at https://github.com/go2hx/go2hx/blob/ee3a8ced204635456c6df3df2c1f67893ff39ad1/stdgo/regexp/syntax/Syntax.hx#L1737

So the effective code is:
```haxe
throw Go.toInterface(Go.recover_exception);
```

Haxe's two errors in the post above are both that a `String` is expected by `throw`.

In Go however the built-in `panic()` function expects an interface to be passed - https://go.dev/ref/spec#Handling_panics.

So in this case (and probably others) Haxe expects a string to be passed, but Go2hx is providing an interface.

I imagine some way needs to be found for the argument to Haxe `throw` to always be cast to Haxe `String`, maybe by calling its `string()` method.

**Comment by elliott5** [comment]
An alternative, better approach, would be to find a way to encode/decode Go2hx interface values to/from Haxe string.

**Comment by PXshadow** [comment]
@elliott5 Yes that's a great point, I'll work on that now. and write here how it goes.

**Comment by PXshadow** [comment]
The invalid string issue has been fixed with a check for string validity before turning into a Haxe string.

The exception issue comes from here:

https://github.com/go2hx/go2hx/blob/5bf82d932f9f68db6c693c1f19796c4e063b1004/stdgo/regexp/syntax/Syntax.hx#LL4882C52-L4882C52

Because this is a test called TestParseInvalidRegexps I assume all of these tests are supposed to fail and perhaps this test is slipping through the try catch somehow but this is just speculation.

If you have any thoughts feel free to share them @elliott5

**Comment by elliott5** [comment]
There is no explicit cast to Haxe `String` in the `throw` statement at the start of the quoted generated code.

Does `AnyInterfaceData` not need `@:to` in front of https://github.com/go2hx/go2hx/blob/5bf82d932f9f68db6c693c1f19796c4e063b1004/stdgo/StdGoTypes.hx#L1731
for an implicit cast to `String` to be known to the compiler? As described in https://haxe.org/manual/types-abstract-implicit-casts.html

**Comment by PXshadow** [comment]
> There is no explicit cast to Haxe `String` in the `throw` statement at the start of the quoted generated code.
> 
> Does `AnyInterfaceData` not need `@:to` in front of
> 
> https://github.com/go2hx/go2hx/blob/5bf82d932f9f68db6c693c1f19796c4e063b1004/stdgo/StdGoTypes.hx#L1731
> 
> for an implicit cast to `String` to be known to the compiler? As described in https://haxe.org/manual/types-abstract-implicit-casts.html

I don't believe Haxe excepts a string for a throw statement/expression, here is an example:  https://try.haxe.org/#7A1a751a

You are right an implicit cast, the type needs to be an abstract and then needs to be specified either with a ``to X`` at the top of the abstract or the ``@:to`` metadata.

**Comment by elliott5** [comment]
Well done for giving an example that proves "String" is not the issue.

My suspicion now turns to the casting of an initialised `GoRefMap` to `GoMap` for `_p.height` in the code you sent me:
```haxe
    static public function _checkHeight( _p:Ref<T_parser>, _re:Ref<Regexp>):Void {
        if (_p._numRegexp < (1000 : GoInt)) {
            return;
        };
        if (_p._height == null) {
            _p._height = ({
                final x = new GoRefMap<Ref<stdgo.regexp.syntax.Syntax.Regexp>, GoInt>();
                @:mergeBlock {};
                cast x;
            } : GoMap<Ref<stdgo.regexp.syntax.Syntax.Regexp>, GoInt>);
            for (__0 => _re in _p._stack) {
                _p._checkHeight(_re);
            };
        };
        if (_p._calcHeight(_re, true) > (1000 : GoInt)) {
            throw Go.toInterface(Go.asInterface((("expression nests too deeply" : GoString) : stdgo.regexp.syntax.Syntax.ErrorCode)));
        };
    }
```

`Ref<>` is pretending a Haxe class is a pointer to that class. My worry is whether the loss of knowledge of that pretence (perhaps by the cast to `GoMap` from `GoRefMap`) means that knowledge of that pretence is being lost somewhere and an infinite loop being created as a result.

**Comment by PXshadow** [comment]
> Well done for giving an example that proves "String" is not the issue.
> 
> My suspicion now turns to the casting of an initialised `GoRefMap` to `GoMap` for `_p.height` in the code you sent me:
> 
> ```hx
>     static public function _checkHeight( _p:Ref<T_parser>, _re:Ref<Regexp>):Void {
>         if (_p._numRegexp < (1000 : GoInt)) {
>             return;
>         };
>         if (_p._height == null) {
>             _p._height = ({
>                 final x = new GoRefMap<Ref<stdgo.regexp.syntax.Syntax.Regexp>, GoInt>();
>                 @:mergeBlock {};
>                 cast x;
>             } : GoMap<Ref<stdgo.regexp.syntax.Syntax.Regexp>, GoInt>);
>             for (__0 => _re in _p._stack) {
>                 _p._checkHeight(_re);
>             };
>         };
>         if (_p._calcHeight(_re, true) > (1000 : GoInt)) {
>             throw Go.toInterface(Go.asInterface((("expression nests too deeply" : GoString) : stdgo.regexp.syntax.Syntax.ErrorCode)));
>         };
>     }
> ```
> 
> `Ref<>` is pretending a Haxe class is a pointer to that class. My worry is whether the loss of knowledge of that pretence somewhere (perhaps by the cast to `GoMap` from `GoRefMap`) means that an infinite loop being created as a result.

That's a really good guess I will reinspect it when we get back there, I have slightly bad news though, after rebuilding the std library and bringing in some new fixes, I believe the others tests where false positive passing and now the current front lines have been moved back to ``TestParseSimple`` the first test suite in ``regexp/syntax`` 

Here is the test failing in a partially isolated format:

```go
package main

import (
	"fmt"
	"regexp/syntax"
	"strings"
	"unicode"
)

func main() {
	r, err := syntax.Parse(`\Q+|*?{[\E`, syntax.Flags(204))
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(dump(r))
	}
}

// dump prints a string representation of the regexp showing
// the structure explicitly.
func dump(re *syntax.Regexp) string {
	var b strings.Builder
	dumpRegexp(&b, re)
	return b.String()
}

var opNames = []string{
	syntax.OpNoMatch:        "no",
	syntax.OpEmptyMatch:     "emp",
	syntax.OpLiteral:        "lit",
	syntax.OpCharClass:      "cc",
	syntax.OpAnyCharNotNL:   "dnl",
	syntax.OpAnyChar:        "dot",
	syntax.OpBeginLine:      "bol",
	syntax.OpEndLine:        "eol",
	syntax.OpBeginText:      "bot",
	syntax.OpEndText:        "eot",
	syntax.OpWordBoundary:   "wb",
	syntax.OpNoWordBoundary: "nwb",
	syntax.OpCapture:        "cap",
	syntax.OpStar:           "star",
	syntax.OpPlus:           "plus",
	syntax.OpQuest:          "que",
	syntax.OpRepeat:         "rep",
	syntax.OpConcat:         "cat",
	syntax.OpAlternate:      "alt",
}

// dumpRegexp writes an encoding of the syntax tree for the regexp re to b.
// It is used during testing to distinguish between parses that might print
// the same using re's String method.
func dumpRegexp(b *strings.Builder, re *syntax.Regexp) {
	if int(re.Op) >= len(opNames) || opNames[re.Op] == "" {
		fmt.Fprintf(b, "op%d", re.Op)
	} else {
		switch re.Op {
		default:
			b.WriteString(opNames[re.Op])
		case syntax.OpStar, syntax.OpPlus, syntax.OpQuest, syntax.OpRepeat:
			if re.Flags&syntax.NonGreedy != 0 {
				b.WriteByte('n')
			}
			b.WriteString(opNames[re.Op])
		case syntax.OpLiteral:
			if len(re.Rune) > 1 {
				b.WriteString("str")
			} else {
				b.WriteString("lit")
			}
			if re.Flags&syntax.FoldCase != 0 {
				for _, r := range re.Rune {
					if unicode.SimpleFold(r) != r {
						b.WriteString("fold")
						break
					}
				}
			}
		}
	}
	b.WriteByte('{')
	switch re.Op {
	case syntax.OpEndText:
		if re.Flags&syntax.WasDollar == 0 {
			b.WriteString(`\z`)
		}
	case syntax.OpLiteral:
		for _, r := range re.Rune {
			b.WriteRune(r)
		}
	case syntax.OpConcat, syntax.OpAlternate:
		for _, sub := range re.Sub {
			dumpRegexp(b, sub)
		}
	case syntax.OpStar, syntax.OpPlus, syntax.OpQuest:
		dumpRegexp(b, re.Sub[0])
	case syntax.OpRepeat:
		fmt.Fprintf(b, "%d,%d ", re.Min, re.Max)
		dumpRegexp(b, re.Sub[0])
	case syntax.OpCapture:
		if re.Name != "" {
			b.WriteString(re.Name)
			b.WriteByte(':')
		}
		dumpRegexp(b, re.Sub[0])
	case syntax.OpCharClass:
		sep := ""
		for i := 0; i < len(re.Rune); i += 2 {
			b.WriteString(sep)
			sep = " "
			lo, hi := re.Rune[i], re.Rune[i+1]
			if lo == hi {
				fmt.Fprintf(b, "%#x", lo)
			} else {
				fmt.Fprintf(b, "%#x-%#x", lo, hi)
			}
		}
	}
	b.WriteByte('}')
}
```

**Comment by PXshadow** [comment]
Regexp/syntax passes all tests, above example code works.

TestBadCompile fails as a signal11 crash on hl. If skipped this is the result:
```
=== RUN  TestGoodCompile
-- PASS: TestGoodCompile (0.0118029117584229)

=== RUN  TestBadCompile
-- PASS: TestBadCompile (5.29289245605469e-05)

=== RUN  TestMatch
-- PASS: TestMatch (0.119991064071655)

=== RUN  TestMatchFunction
-- PASS: TestMatchFunction (0.0959568023681641)

=== RUN  TestCopyMatch
-- PASS: TestCopyMatch (0.123787879943848)

=== RUN  TestReplaceAll
-- PASS: TestReplaceAll (0.198426008224487)

=== RUN  TestReplaceAllLiteral
-- PASS: TestReplaceAllLiteral (0.136915922164917)

=== RUN  TestReplaceAllFunc
-- PASS: TestReplaceAllFunc (0.0138318538665771)

=== RUN  TestQuoteMeta
QuoteMeta(`日本語+`) = `日本語+`; want `日本語\+`
```

**Comment by elliott5** [comment]
```
QuoteMeta(`日本語+`) = `日本語+`; want `日本語\+`
```

May be a fault in the way that the (un-)quoting works, rather than in `regexp` itself.

**Comment by PXshadow** [comment]
Other stdlibs to focus on first. Another issue can be made if needed at a later time (same reason as #153)
